crates/red_knot_python_semantic/src/semantic_index/ast_ids.rs

```diff
@@ -110,38 +146,49 @@ impl HasScopedAstId for ast::ExpressionRef<'_> {
 
 #[derive(Debug)]
 pub(super) struct AstIdsBuilder {
-    next_id: ScopedExpressionId,
+    next_expression_id: ScopedExpressionId,
+    next_use_id: ScopedUseId,
     expressions_map: FxHashMap<ExpressionNodeKey, ScopedExpressionId>,
+    uses_map: FxHashMap<ExpressionNodeKey, ScopedUseId>,
 }
 
 impl AstIdsBuilder {
     pub(super) fn new() -> Self {
         Self {
-            next_id: ScopedExpressionId::new(0),
+            next_expression_id: ScopedExpressionId::new(0),
+            next_use_id: ScopedUseId::new(0),
             expressions_map: FxHashMap::default(),
+            uses_map: FxHashMap::default(),
         }
     }
 
-    /// Adds `expr` to the AST ids map and returns its id.
-    ///
-    /// ## Safety
-    /// The function is marked as unsafe because it calls [`AstNodeRef::new`] which requires
-    /// that `expr` is a child of `parsed`.
-    #[allow(unsafe_code)]
+    /// Adds `expr` to the expression ids map and returns its id.
     pub(super) fn record_expression(&mut self, expr: &ast::Expr) -> ScopedExpressionId {
-        let expression_id = self.next_id;
-        self.next_id = expression_id + 1;
+        let expression_id = self.next_expression_id;
+        self.next_expression_id = expression_id + 1;
 
         self.expressions_map.insert(expr.into(), expression_id);
 
         expression_id
     }
 
+    /// Adds `expr` to the use ids map and returns its id.
+    pub(super) fn record_use(&mut self, expr: &ast::Expr) -> ScopedUseId {
+        let use_id = self.next_use_id;
+        self.next_use_id = use_id + 1;
+
+        self.uses_map.insert(expr.into(), use_id);
```

Comment:
I know you just copied what I did, so this is really my mistake. Can't we use `self.use_map.len()` to compute the next id?


```suggestion
        let use_id = NextUseId::from_usize(self.uses_map.len());
        self.uses_map.insert(expr.into(), use_id);
```